flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java

```diff
@@ -661,6 +663,56 @@ protected void reportOrForwardLatencyMarker(LatencyMarker marker) {
 
 	// ----------------------- Helper classes -----------------------
 
+	/**
+	 * Wrapping {@link Output} that adds an emission timestamp to records.
+	 */
+	public static class EmissionTimestamperOutput<OUT> implements Output<StreamRecord<OUT>> {
+		private final Output<StreamRecord<OUT>> output;
+		private final String operatorId;
+
+		public EmissionTimestamperOutput(Output<StreamRecord<OUT>> output, Counter counter, String operatorId) {
+			LOG.info("CREATED INSTANCE OF EMISSION TIMESTAMPER");
+			this.output = new CountingOutput<>(output, counter);
+			this.operatorId = operatorId;
+		}
+		public EmissionTimestamperOutput(Output<StreamRecord<OUT>> output, String operatorId) {
+			LOG.info("CREATED INSTANCE OF EMISSION TIMESTAMPER");
+			this.output = output;
+			this.operatorId = operatorId;
+		}
+
+		@Override
+		public void emitWatermark(Watermark mark) {
+			output.emitWatermark(mark);
+		}
+
+		@Override
+		public void emitLatencyMarker(LatencyMarker latencyMarker) {
+			output.emitLatencyMarker(latencyMarker);
+		}
+
+		@Override
+		public void collect(StreamRecord<OUT> record) {
+			LOG.info("Add ts to record: (" + operatorId+ ", "+ System.currentTimeMillis() + ") with " + record.getOperatorOutputTimestamps().size() + "elements");
```

Comment:
If these two loging statements are purely for logging, I would downgrade them to debug level.
They will affect throughput otherwise.